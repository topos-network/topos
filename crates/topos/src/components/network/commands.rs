use clap::{Args, Subcommand};

mod spam;

pub(crate) use spam::Spam;

#[derive(Args, Debug)]
pub(crate) struct NetworkCommand {
    #[clap(subcommand)]
    pub(crate) subcommands: Option<NetworkCommands>,
}

#[derive(Subcommand, Debug)]
pub(crate) enum NetworkCommands {
    Spam(Box<Spam>),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_run() {
        assert!(NetworkCommands::has_subcommand("spam"));
    }
}
