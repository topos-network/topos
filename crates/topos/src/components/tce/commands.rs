use std::path::PathBuf;

use clap::{Args, Subcommand};

mod peer_id;
mod push_certificate;
mod run;
mod status;

pub(crate) use push_certificate::PushCertificate;
pub(crate) use run::Run;
pub(crate) use status::Status;

use self::peer_id::Keys;

#[derive(Args, Debug)]
pub(crate) struct NodeArgument {
    #[clap(short, long, default_value = "http://[::1]:1340")]
    pub(crate) node: String,
}

/// Topos CLI subcommand for the TCE related functionalities
#[derive(Args, Debug)]
pub(crate) struct TceCommand {
    #[clap(from_global)]
    pub(crate) verbose: u8,

    #[clap(from_global)]
    pub(crate) home: PathBuf,

    #[clap(subcommand)]
    pub(crate) subcommands: Option<TceCommands>,
}

#[derive(Subcommand, Debug)]
pub(crate) enum TceCommands {
    PushCertificate(PushCertificate),
    Keys(Keys),
    Run(Box<Run>),
    Status(Status),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_run() {
        assert!(TceCommands::has_subcommand("run"));
    }
}
