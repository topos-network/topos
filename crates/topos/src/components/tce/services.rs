pub(crate) mod push_certificate;
mod status;
