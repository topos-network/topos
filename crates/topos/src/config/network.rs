use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct NetworkConfig {
    /// The target node api endpoint.
    /// Multiple nodes could be specified as comma separated list
    #[serde(default = "default_target_nodes")]
    pub target_nodes: Vec<String>,

    /// Path to json file with list of target nodes as alternative to `target_nodes`
    pub target_nodes_path: Option<String>,

    /// Seed for generation of local private signing keys and corresponding subnet ids.
    #[serde(default = "default_local_key_seed")]
    pub local_key_seed: u64,

    /// Certificates generated in one batch. Batch is generated every `batch-interval` milliseconds.
    #[serde(default = "default_cert_per_batch")]
    pub cert_per_batch: u8,

    /// Number of subnets to use for certificate generation. For every certificate subnet id will be picked randomly.
    #[serde(default = "default_nb_subnets")]
    pub nb_subnets: u8,

    /// Number of batches to generate before finishing execution.
    /// If not specified, batches will be generated indefinitely.
    #[serde(default = "default_nb_batches")]
    pub nb_batches: u64,

    /// Time interval in milliseconds between generated batches of certificates
    #[serde(default = "default_batch_interval")]
    pub batch_interval: u64,

    /// List of generated certificate target subnets. No target subnets by default.
    #[serde(default = "default_target_subnets")]
    pub target_subnets: Vec<String>,

    /// Socket of the opentelemetry agent endpoint.
    /// If not provided open telemetry will not be used
    pub otlp_agent: Option<String>,

    /// Otlp service name.
    /// If not provided open telemetry will not be used
    pub otlp_service_name: Option<String>,
}

fn default_target_nodes() -> Vec<String> {
    vec!["http://[::1]:1340,http://[::1]:1341".to_string()]
}

fn default_local_key_seed() -> u64 {
    1
}

fn default_cert_per_batch() -> u8 {
    1
}

fn default_nb_subnets() -> u8 {
    1
}

fn default_nb_batches() -> u64 {
    10
}

fn default_batch_interval() -> u64 {
    2000
}

fn default_target_subnets() -> Vec<String> {
    vec![
        "0x3bc19e36ff1673910575b6727a974a9abd80c9a875d41ab3e2648dbfb9e4b518,0xa00d60b2b408c2a14c5d70cdd2c205db8985ef737a7e55ad20ea32cc9e7c417c".to_string()
    ]
}

impl Default for NetworkConfig {
    fn default() -> Self {
        NetworkConfig {
            target_nodes: default_target_nodes(),
            target_nodes_path: None,
            local_key_seed: default_local_key_seed(),
            cert_per_batch: default_cert_per_batch(),
            nb_subnets: default_nb_subnets(),
            nb_batches: default_nb_batches(),
            batch_interval: default_batch_interval(),
            target_subnets: default_target_subnets(),
            otlp_agent: None,
            otlp_service_name: None,
        }
    }
}
