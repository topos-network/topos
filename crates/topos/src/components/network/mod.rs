use self::commands::{NetworkCommand, NetworkCommands};
use tokio::{signal, spawn};
use topos_certificate_spammer::CertificateSpammerConfiguration;
use tracing::{error, info};

use crate::tracing::setup_tracing;

pub(crate) mod commands;

pub(crate) async fn handle_command(
    NetworkCommand {
        subcommands,
        verbose,
    }: NetworkCommand,
) -> Result<(), Box<dyn std::error::Error>> {
    match subcommands {
        Some(NetworkCommands::Spam(cmd)) => {
            let config = CertificateSpammerConfiguration {
                target_node: cmd.target_node,
                target_nodes_path: cmd.target_nodes_path,
                cert_per_batch: cmd.cert_per_batch,
                batch_time_interval: cmd.batch_time_interval,
                nb_subnets: cmd.nb_subnets,
                byzantine_threshold: cmd.byzantine_threshold,
                node_per_cert: cmd.node_per_cert,
            };

            // Setup instrumentation if both otlp agent and otlp service name
            // are provided as arguments
            // We may want to use instrumentation in e2e tests
            setup_tracing(verbose, cmd.otlp_agent, cmd.otlp_service_name)?;

            spawn(async move {
                if let Err(error) = topos_certificate_spammer::run(config).await {
                    panic!("Unable to execute network spam command due to : {error:?}");
                }
            });

            signal::ctrl_c()
                .await
                .expect("failed to listen for signals");

            Ok(())
        }
        None => Ok(()),
    }
}
