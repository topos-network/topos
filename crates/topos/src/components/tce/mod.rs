use std::{
    fs::File,
    io::{self, Read},
    path::Path,
    str::FromStr,
    sync::Arc,
};

use tokio::{signal, spawn, sync::Mutex};
use tonic::transport::Channel;
use topos_core::api::tce::v1::console_service_client::ConsoleServiceClient;
use topos_p2p::PeerId;
use topos_tce::TceConfiguration;
use tower::Service;
use tracing::{debug, trace};

use crate::options::input_format::{InputFormat, Parser};

use self::commands::{TceCommand, TceCommands};

pub(crate) mod commands;
pub(crate) mod services;

pub(crate) struct TCEService {
    pub(crate) client: Arc<Mutex<ConsoleServiceClient<Channel>>>,
}

pub(crate) struct PeerList(pub(crate) Option<String>);

impl Parser<PeerList> for InputFormat {
    type Result = Result<Vec<PeerId>, io::Error>;

    fn parse(&self, PeerList(input): PeerList) -> Self::Result {
        let mut input_string = String::new();
        _ = match input {
            Some(path) if Path::new(&path).is_file() => {
                File::open(path)?.read_to_string(&mut input_string)?
            }
            Some(string) => {
                input_string = string;
                0
            }
            None => io::stdin().read_to_string(&mut input_string)?,
        };

        match self {
            InputFormat::Json => Ok(serde_json::from_str::<Vec<PeerId>>(&input_string)?),
            InputFormat::Plain => Ok(input_string
                .trim()
                .split(&[',', '\n'])
                .filter_map(|s| PeerId::from_str(s.trim()).ok())
                .collect()),
        }
    }
}

pub(crate) async fn handle_command(
    TceCommand {
        mut endpoint,
        subcommands,
    }: TceCommand,
) -> Result<(), Box<dyn std::error::Error>> {
    match subcommands {
        Some(TceCommands::PushPeerList(cmd)) => {
            debug!("Start executing PushPeerList command");
            trace!("Building the gRPC client with {:?}", endpoint);

            let client = Arc::new(Mutex::new(
                ConsoleServiceClient::connect(endpoint.take().unwrap())
                    .await
                    .unwrap(),
            ));
            trace!("gRPC client successfully built");

            let mut tce_service = TCEService {
                client: client.clone(),
            };

            debug!("Executing the PushPeerList on the TCE service");
            tce_service.call(cmd).await?;

            Ok(())
        }

        Some(TceCommands::Run(cmd)) => {
            let config = TceConfiguration {
                boot_peers: cmd.parse_boot_peers(),
                db_path: cmd.db_path,
                local_key_seed: cmd.local_key_seed,
                jaeger_agent: cmd.jaeger_agent,
                jaeger_service_name: cmd.jaeger_service_name,
                tce_local_port: cmd.tce_local_port,
                trbp_params: cmd.trbp_params,
                api_addr: cmd.api_addr,
            };

            spawn(async move {
                topos_tce::run(&config).await;
            });

            signal::ctrl_c()
                .await
                .expect("failed to listen for signals");

            Ok(())
        }
        None => Ok(()),
    }
}
