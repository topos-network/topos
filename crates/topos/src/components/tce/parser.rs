use std::{
    fs::File,
    io::{self, Read},
    path::{Path, PathBuf},
    str::FromStr,
    sync::Arc,
    time::Duration,
};

use opentelemetry::global;
use tokio::{
    signal, spawn,
    sync::{mpsc, oneshot, Mutex},
};
use tonic::transport::Channel;
use topos_core::api::grpc::tce::v1::console_service_client::ConsoleServiceClient;
use topos_p2p::PeerId;
use topos_tce::config::{StorageConfiguration, TceConfiguration};
use tower::Service;
use tracing::{debug, error, info, trace};

use crate::options::input_format::{InputFormat, Parser};
use crate::tracing::setup_tracing;

pub(crate) struct PeerList(pub(crate) Option<String>);

impl Parser<PeerList> for InputFormat {
    type Result = Result<Vec<PeerId>, io::Error>;

    fn parse(&self, PeerList(input): PeerList) -> Self::Result {
        let mut input_string = String::new();
        _ = match input {
            Some(path) if Path::new(&path).is_file() => {
                File::open(path)?.read_to_string(&mut input_string)?
            }
            Some(string) => {
                input_string = string;
                0
            }
            None => io::stdin().read_to_string(&mut input_string)?,
        };

        match self {
            InputFormat::Json => Ok(serde_json::from_str::<Vec<PeerId>>(&input_string)?),
            InputFormat::Plain => Ok(input_string
                .trim()
                .split(&[',', '\n'])
                .filter_map(|s| PeerId::from_str(s.trim()).ok())
                .collect()),
        }
    }
}
