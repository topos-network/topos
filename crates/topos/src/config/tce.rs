use std::path::Path;
use std::{net::SocketAddr, path::PathBuf};

use figment::{
    providers::{Format, Serialized, Toml},
    Figment,
};
use serde::{Deserialize, Serialize};

use crate::components::tce::commands::Run;
use crate::config::Config;
use topos_p2p::{Multiaddr, PeerId};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct TceConfig {
    /// Storage database path, if not set RAM storage is used
    #[serde(default = "default_db_path")]
    pub db_path: PathBuf,
    /// Array of extra boot nodes to connect to
    pub extra_boot_peers: Option<String>,
    /// Ip for the p2p Multiaddr
    pub tce_ext_host: Option<String>,
    /// Port for the p2p Multiaddr
    pub tce_local_port: Option<u16>,
    /// Local peer secret key seed (optional, used for testing)
    pub local_key_seed: Option<String>,
    /// Connection degree for the GossipSub overlay
    pub minimum_tce_cluster_size: Option<usize>,
    /// gRPC API Addr
    #[serde(default = "default_grpc_api_addr")]
    pub grpc_api_addr: SocketAddr,
    /// GraphQL API Addr
    #[serde(default = "default_graphql_api_addr")]
    pub graphql_api_addr: SocketAddr,
    /// Metrics server API Addr
    #[serde(default = "default_metrics_api_addr")]
    pub metrics_api_addr: SocketAddr,
    /// Socket of the opentelemetry agent endpoint
    /// If not provided open telemetry will not be used
    pub otlp_agent: Option<String>,
    /// Otlp service name
    /// If not provided open telemetry will not be used
    pub otlp_service_name: Option<String>,
}

fn default_db_path() -> PathBuf {
    PathBuf::from("./tce_rocksdb")
}

fn default_grpc_api_addr() -> SocketAddr {
    "0.0.0.0:1340"
        .parse()
        .expect("Cannot parse address to SocketAddr")
}

fn default_graphql_api_addr() -> SocketAddr {
    "0.0.0.0:4030"
        .parse()
        .expect("Cannot parse address to SocketAddr")
}

fn default_metrics_api_addr() -> SocketAddr {
    "0.0.0.0:3000"
        .parse()
        .expect("Cannot parse address to SocketAddr")
}

impl TceConfig {
    pub fn parse_boot_peers(&self) -> Vec<(PeerId, Multiaddr)> {
        self.extra_boot_peers
            .clone()
            .unwrap_or_default()
            .split(&[',', ' '])
            .map(|s| s.to_string())
            .collect::<Vec<String>>()
            .chunks(2)
            .filter_map(|pair| {
                if pair.len() > 1 {
                    Some((
                        pair[0].as_str().parse().unwrap(),
                        pair[1].as_str().parse().unwrap(),
                    ))
                } else {
                    None
                }
            })
            .collect()
    }
}

impl Config for TceConfig {
    type Command = Run;
    type Output = TceConfig;

    fn profile(&self) -> String {
        "tce".to_string()
    }

    fn load_from_file(figment: Figment, home: &Path) -> Figment {
        let home = home.join("config.toml");

        let tce = Figment::new()
            .merge(Toml::file(home).nested())
            .select("tce");

        figment.merge(tce)
    }

    fn load_context(figment: Figment) -> Result<Self::Output, figment::Error> {
        figment.extract()
    }
}
