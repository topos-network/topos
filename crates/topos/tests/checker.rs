use crate::support::network::create_network;
use assert_cmd::{prelude::*, Command};
use std::{net::UdpSocket, thread, time::Duration};
use test_log::test;
use tonic::{Request, Response, Status};
use topos_core::api::tce::v1::{
    console_service_server::{ConsoleService, ConsoleServiceServer},
    PushPeerListRequest, PushPeerListResponse, StatusRequest, StatusResponse,
};

mod support {
    pub mod network;
}

#[test]
fn help_display() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("checker").arg("tce").arg("-h");

    let output = cmd.assert().success();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    insta::assert_snapshot!(result);

    Ok(())
}

#[test]
fn assert_delivery_help_display() -> Result<(), Box<dyn std::error::Error>> {
    let mut cmd = Command::cargo_bin("topos")?;
    cmd.arg("checker")
        .arg("tce")
        .arg("assert-delivery")
        .arg("-h");

    let output = cmd.assert().success();

    let result: &str = std::str::from_utf8(&output.get_output().stdout)?;

    insta::assert_snapshot!(result);

    Ok(())
}

#[test_log::test(tokio::test)]
// FIXME: This test is flaky, it fails sometimes because of sample failure
async fn assert_delivery() -> Result<(), Box<dyn std::error::Error>> {
    let mut peers_context = create_network(10, 4).await;

    let mut status: Vec<bool> = Vec::new();

    for (_peer_id, client) in peers_context.iter_mut() {
        let response = client
            .console_grpc_client
            .status(StatusRequest {})
            .await
            .expect("Can't get status");

        status.push(response.into_inner().has_active_sample);
    }

    assert!(status.iter().all(|s| *s));

    let nodes: String = peers_context
        .iter()
        .map(|peer| peer.1.api_entrypoint.clone())
        .collect::<Vec<_>>()
        .join(",");

    let (tx, rx) = tokio::sync::oneshot::channel::<()>();

    let join = thread::spawn(|| {
        let mut cmd = Command::cargo_bin("topos").unwrap();
        cmd.env("TOPOS_LOG_FORMAT", "json");
        cmd.env("RUST_LOG", "topos=debug");

        cmd.arg("checker")
            .arg("tce")
            .arg("assert-delivery")
            // .args(&["--timeout-broadcast", "15"])
            .args(&["-f", "plain"])
            .arg(nodes);

        let output = cmd.assert().success();

        tx.send(()).unwrap();
    });

    rx.await.unwrap();

    Ok(())
}
