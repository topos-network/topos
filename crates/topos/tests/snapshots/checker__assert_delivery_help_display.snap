---
source: crates/topos/tests/checker.rs
expression: result
---
Assert that a delivery was made to a list of nodes

Usage: topos checker tce assert-delivery [OPTIONS] [NODES]

Arguments:
  [NODES]  The node list to be used, can be a file path or a comma separated list of Uri. If not provided, stdin is listened [env: TARGET_NODES_PATH=]

Options:
  -f, --format <FORMAT>
          [default: json] [possible values: json, plain]
  -v, --verbose...
          
  -t, --timeout <TIMEOUT>
          Global timeout for the command [default: 10]
      --timeout-broadcast <TIMEOUT_BROADCAST>
          Seconds to wait before asserting the broadcast [default: 2]
  -h, --help
          Print help

